[Console] Handle encoded characters in API requests

mibragimov opened the pull request:
Fixes https://github.com/elastic/kibana/issues/131798
Replaces https://github.com/elastic/kibana/pull/132191
Related to https://github.com/elastic/kibana/issues/76585
Related to https://github.com/elastic/kibana/issues/128701

## Summary

Fixes sending API requests with encoded characters in request URLs. 
Fixes the issue when using `_cat` commands with parameters.

### Testing
To test this out, send any of the following requests in Console


```
POST _bulk
{ "index" : { "_index" : "my-index", "_id" : "this/is/a/doc" } }
{ "title" : "a doc" }

GET my-index/_doc/this%2Fis%2Fa%2Fdoc
```
<img width="1118" alt="Screen Shot 2022-06-29 at 19 39 10" src="https://user-images.githubusercontent.com/53621505/176464908-8a715787-b94a-423b-a700-7d6e1690ca1f.png">

```
GET _cat/indices/.kibana?v
```
<img width="1118" alt="Screen Shot 2022-06-29 at 19 40 00" src="https://user-images.githubusercontent.com/53621505/176464969-d0cd942c-939d-4eb2-b849-829d9879e425.png">

```
PUT %{[@metadata][beat]}-%{[@metadata][version]}-2020.08.23
```
<img width="1118" alt="Screen Shot 2022-06-29 at 19 39 33" src="https://user-images.githubusercontent.com/53621505/176465022-66c957b1-66b3-4c2a-8835-fa497bd1ebb7.png">

```
PUT /%3Cmy-index-%7Bnow%2Fd%7D%3E
```
<img width="1118" alt="Screen Shot 2022-06-29 at 19 39 51" src="https://user-images.githubusercontent.com/53621505/176465066-6ec44234-5169-4092-a4cd-acbb14b8821c.png">

### Release note

Fixed a bug in Console when sending a request with encoded characters resulted in an error

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

mibragimov commented:
@elasticmachine merge upstream

mibragimov commented:
@elasticmachine merge upstream

mibragimov commented:
@elasticmachine merge upstream

cjcenizal commented on src/plugins/console/server/lib/utils/index.ts:
```diff
@@ -0,0 +1,28 @@
+import { URLSearchParams } from 'url';
+import { trimStart } from 'lodash';
+
+export const encodePath = (path: string) => {
```
Could you extract this into a `encode_path.ts` file, and re-export `encodePath` from this file? Then could also rename `utils.test.ts` to `encode_path.test.ts`? I think this will make it easier to navigate the codebase by scanning files and being able to understand their role without having to open them individually. It will also scale better as we add more utils.

cjcenizal reviewed (commented):
Thanks for working on this! Love all of the tests. Had a couple comments.

alisonelizabeth commented on src/plugins/console/server/lib/proxy_request.test.ts:
```diff
@@ -63,14 +78,7 @@ describe(`Console's send request`, () => {
-      timeout: 30000,
-      uri: new URL('http://noone.nowhere.none'),
-    });
+    const result1 = await sendProxyRequest({});
```
[nit] Clarify naming
```suggestion
    const defaultResult = await sendProxyRequest({});
```

alisonelizabeth commented on src/plugins/console/server/lib/proxy_request.test.ts:
```diff
@@ -83,14 +91,7 @@ describe(`Console's send request`, () => {
-      timeout: 30000,
-      uri: new URL('http://noone.nowhere.none'),
-    });
+    const result2 = await sendProxyRequest({ headers: { Host: 'myhost' } });
```
[nit]: Clarify naming
```suggestion
    const resultWithHostHeader = await sendProxyRequest({ headers: { Host: 'myhost' } });
```

alisonelizabeth commented on src/plugins/console/server/lib/proxy_request.test.ts:
```diff
@@ -115,39 +116,33 @@ describe(`Console's send request`, () => {
-        method: 'get',
-        payload: null as any,
-        timeout: 30000,
+    const runEncodePathTests = async (initialPath: string, expectedPath: string, uri?: URL) => {
```
[nit] I think this util could be used regardless if the path was encoded. You might consider renaming it to something like: `verifyRequestPath()`

alisonelizabeth commented on src/plugins/console/server/lib/proxy_request.test.ts:
```diff
@@ -115,39 +116,33 @@ describe(`Console's send request`, () => {
+    };
+
+    it('should correctly encode invalid URL characters included in path', async () => {
+      await runEncodePathTests(
```
What do you think about updating this util to accept an object? It might be easier to scan in the future which args represent the input vs. expected result. For example:

```
 await runEncodePathTests({
        initialPath: '_index/.test',
        uri: new URL('http://noone.nowhere.none/_index/.test?q=something&v=something'),
        expectedPath: '_index/.test?q=something&v=something',
      });
```

alisonelizabeth commented on src/plugins/console/server/lib/utils/encode_path.ts:
```diff
@@ -0,0 +1,28 @@
+import { trimStart } from 'lodash';
+
+export const encodePath = (path: string) => {
+  const decodedPath = new URLSearchParams(`path=${path}`).get('path') ?? '';
```
Q: Is the `??` necessary - will this ever evaluate to `null` if we're passing `path=` and the path value is also required?

mibragimov replied:
`??` is actually not necessary, since we are initializing the path with `path=`, this will evaluate to `null` only if the path value is `null` which won't happen because the path value is always string and required in this case. `?? ''` is used to avoid the ts error (`TS2345: Argument of type 'string | null' is not assignable to parameter of type 'string | undefined'.`) on the following line.

```
const isEncoded = trimStart(path, '/') !== trimStart(decodedPath, '/');
```
 Perhaps it'd be better to get around this with an if statement. WDYT?

```
 let isEncoded;
 if (decodedPath) {
    isEncoded = trimStart(path, '/') !== trimStart(decodedPath, '/');
 }
```

alisonelizabeth replied:
Ah, I see. I'm OK to leave as-is. You could also do:

```
const isEncoded = trimStart(path, '/') !== trimStart(decodedPath as string, '/');
```

alisonelizabeth reviewed (commented):
Great job with this @mibragimov! Verified locally. I left a couple comments in the code, mostly around naming/improving test readability. Let me know what you think.

Also, since this is a known bug, I think the label should be updated to `release_note:fix` and we should add a small release note to the PR description. Let me know if you need any help with this. Thanks!

mibragimov commented:
Thank you for the review @alisonelizabeth. I've addressed your feedback with https://github.com/elastic/kibana/pull/135441/commits/cbe649594db93ca630444f91a8efefc346d74cf0. Do you mind having another look? Thanks!

alisonelizabeth reviewed (approved):
Thanks for addressing my feedback! Latest LGTM.

cjcenizal reviewed (changes requested):
Is it possible to add a functional test to verify that requests with query parameters behave as expected? This will guard against regressions like the one added in https://github.com/elastic/kibana/pull/132191.

mibragimov commented:
> Is it possible to add a functional test to verify that requests with query parameters behave as expected? This will guard against regressions like the one added in #132191.

Sure, no problem. I'll get on to it

mibragimov commented:
@elasticmachine merge upstream

mibragimov commented:
@elasticmachine merge upstream

cjcenizal reviewed (approved):
Functional test looks great! Thank you!
